src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -68,145 +70,7 @@ private int CopyReadSegment(ArraySegment<byte> readSegment, ArraySegment<byte> b
-
-        private async Task CopyToAsyncPeekAwaited(
-            ValueTask<ArraySegment<byte>> peekTask,
-            Stream destination,
-            CancellationToken cancellationToken = default(CancellationToken))
-        {
-            while (true)
-            {
-                var segment = await peekTask;
-
-                if (segment.Count == 0)
-                {
-                    return;
-                }
-
-                try
-                {
-                    await destination.WriteAsync(segment.Array, segment.Offset, segment.Count, cancellationToken);
-                }
-                finally
-                {
-                    ConsumedBytes(segment.Count);
-                }
-
-                peekTask = PeekAsync(cancellationToken);
-            }
-        }
-
-        private async Task CopyToAsyncDestinationAwaited(
-            Task destinationTask,
-            int bytesConsumed,
-            Stream destination,
-            CancellationToken cancellationToken = default(CancellationToken))
-        {
-            try
-            {
-                await destinationTask;
-            }
-            finally
-            {
-                ConsumedBytes(bytesConsumed);
-            }
-
-            var peekTask = PeekAsync(cancellationToken);
-
-            if (!peekTask.IsCompleted)
-            {
-                TryProduceContinue();
-            }
-
-            await CopyToAsyncPeekAwaited(peekTask, destination, cancellationToken);
-        }
-
-        public Task Consume(CancellationToken cancellationToken = default(CancellationToken))
-        {
-            while (true)
-            {
-                var task = PeekAsync(cancellationToken);
-                if (!task.IsCompleted)
-                {
-                    TryProduceContinue();
-
-                    // Incomplete Task await result
-                    return ConsumeAwaited(task, cancellationToken);
-                }
-                else
-                {
-                    // ValueTask uses .GetAwaiter().GetResult() if necessary
-                    if (task.Result.Count == 0)
-                    {
-                        // Completed Task, end of stream
-                        return TaskCache.CompletedTask;
-                    }
-
-                    ConsumedBytes(task.Result.Count);
-                }
-            }
-        }
-
-        private async Task ConsumeAwaited(ValueTask<ArraySegment<byte>> currentTask, CancellationToken cancellationToken)
-        {
-            while (true)
-            {
-                var count = (await currentTask).Count;
-
-                if (count == 0)
-                {
-                    // Completed Task, end of stream
-                    return;
-                }
-
-                ConsumedBytes(count);
-                currentTask = PeekAsync(cancellationToken);
-            }
-        }
-
-        private void TryProduceContinue()
+        public void TryProduceContinue()
```

Comment:
Why did you change this from private to public? This is only being called by MessageBody.ReadAsync(), right?